### ISSUE TITLE
How to open database with immutable flag?

### ISSUE THREAD

**Issue by JanValiska** [kkharji/sqlite.lua#131]
Hi,

this is more question than bug report.
I need to read firefox's bookmark database. But when FF is started the database is locked.

```
➜  hbo9xv4w.default sqlite3 places.sqlite
SQLite version 3.37.0 2021-11-27 14:13:22
Enter ".help" for usage hints.
sqlite> .tables
Error: database is locked
sqlite> 
```

When I open  database with immutable flag, I can read items from database:
```
➜  hbo9xv4w.default sqlite3 'file:places.sqlite?immutable=1'
SQLite version 3.37.0 2021-11-27 14:13:22
Enter ".help" for usage hints.
sqlite> .tables
moz_anno_attributes                    
moz_annos                              
moz_bookmarks 
...
```

Is it possible to open database using `sqlite.lua` library in immutable manner? 

Thank you.

**Comment by tami5** [comment]
hmmm I believe is either append `?immutable=1` or a support for `sqlite3_open_v2` is needed for this to work. could you try the first solution and how did you install sqlite.lua? using luarocks?

**Comment by JanValiska** [comment]
I use sqlite.lua in neovim text editor using `packer` plugin manager. Plugin is basically cloned into plugins directory.

I tried append `?immutable=1` to database file path provided to `sqlite.new(path):open()` function without luck.
Perhaps the `sqlite3_open_v2` support is needed as you wrote.
Can I somehow help here?

**Comment by tami5** [comment]
I just did some reasearch and it seems to me that it's mostly likely something that needs to be appended to the uri, see [Recognized Query Parameters](https://sqlite.org/uri.html) and [URI filename examples](https://sqlite.org/c3ref/open.html#urifilenamesinsqlite3open) section. 

I'm starting to believe that maybe open_v2 won't be needed and perhaps some internal url processing strips the query params.

>  Can I somehow help here?
Can you navigate to sqlite/defs, and run the following throw evaluating the file

```lua 
  local uri = 'file:places.sqlite?immutable=1'
  local opts = {}
  local conn = M.get_new_db_ptr()
  local code = clib.sqlite3_open(uri, conn)

  if code ~= M.flags.ok then
    error(("sqlite.lua: couldn't connect to sql database, ERR:"):format(code))
  end

  for k, v in pairs(opts) do
    if not M.valid_pargma[k] then
      error("sqlite.lua: " .. k .. " is not a valid pragma")
    end
    if type(k) == "boolean" then
      k = "ON"
    end
    M.exec_stmt(conn[0], ("pragma %s = %s"):format(k, v))
  end

  vim.inspect(conn[0])
```

btw defs.lua is where ffi interface are. so if changes needs to made it will be made here.

**Comment by JanValiska** [comment]
I tried to evaluate code you posted above. But I changed the uri to contain absolute path to sqlite db like:

```
local uri = 'file:/home/janko/.mozilla/firefox/hbo9xv4w.default/places.sqlite?immutable=1'
```

Evaluation ended up on
 ```
defs.lua:690: sqlite.lua: couldn't connect to sql database, ERR:
```

**Comment by tami5** [comment]
Okay @Conni2461, informed me that perhaps open_uri need to be passed for this to work. Can you change the value of code to `clib.sqlite3_open_v2(uri, conn, M.flags.open_uri, nil)`. 

Also, to see errors change the body of `code ~= M.flags.ok` to 

```lua

M.get_err_msg = function(code)
  for msg, ecode in pairs(M.flags) do
    if code == ecode then
      return msg
    end
  end
end

  if code ~= M.flags.ok then
    error(("sqlite.lua: couldn't connect to sql database, ERR: %s"):format(M.get_err_msg(code)))
  end


```

**Comment by tami5** [comment]
@JanValiska can u checkout #132 and pass as second argument `{open_mode = "ro"}`? https://github.com/dhruvmanila/telescope-bookmarks.nvim/blob/main/lua/telescope/_extensions/bookmarks/firefox.lua#L72. I believe it should work

**Comment by JanValiska** [comment]
Chcecked out #132 and added second agrument to `new` function on L32 of `firefox.lua`. I got this error:

```
E5108: Error executing lua ...nvim/site/pack/packer/opt/sqlite.lua/lua/sqlite/defs.lua:699: missing declaration for symbol 'sqlite3_eval'                                                                          
stack traceback:                                                                                                                                                                                                   
        [C]: in function '__index'                                                                                                                                                                                 
        ...nvim/site/pack/packer/opt/sqlite.lua/lua/sqlite/defs.lua:699: in function '__index'   
```

**Comment by tami5** [comment]
That's odd. why the heck it's calling eval. are you still calling eval somewhere?

**Comment by tami5** [comment]
oh wait the uri needs to be edited and have `?immutable=1` as well. I'm testing telescope-bookmarks right now

**Comment by JanValiska** [comment]
I've changed db connection to:
```
  local db = sqlite.new(utils.join_path("file:"..profile_dir, "places.sqlite?immutable=1"),{open_mode = "ro"}):open()
```
and it works.

Also as you found I forgot to remove old eval functon from `collect_bookmarks` function :/

**Comment by tami5** [comment]
try 
```lua
  local uri = "file:" .. utils.join_path(profile_dir, "places.sqlite") .. "?immutable=1"
  local db = sqlite.new(uri, { open_mode = "ro" }):open()
```

**Comment by tami5** [comment]
> ```
>   local db = sqlite.new(utils.join_path("file:"..profile_dir, "places.sqlite?immutable=1"),{open_mode = "ro"}):open()
> ```

Nice, as soon as I merge open_v2 pr, another pr could be open in telescope-bookmarks to somehow have the above changes. For now, just have fun  with telescope-bookmarks

**Comment by JanValiska** [comment]
Awesome! Thank you! :)

**Comment by tami5** [comment]
This should be closed by #132 merge automatically, np you are welcome.
